### PR TITLE
feat(openebs-upgrade) : Initial check in for upgrade result CR

### PR DESCRIPTION
**Pull request by sagarkrsd**
**What this PR does / why we need it**:
This PR will introduce the schema and auto-generated code for upgrade result CR which will contain the results of all the tasks for upgrading a particular openebs component such as some pv or sp.
This CR can be consumed by other tools to maintain resiliency or build workflow for upgrades.

An example yaml could look like :
```yaml
apiVersion: "openebs.io/v1alpha1"
kind: UpgradeResult
metadata:
  name: pv1-upgrade-result
  namespace: default
  ownerReference: pvc-1
config: [Add task-executor config here]
tasks:
- name: patch-target-deploy
  status: completed
  lastTransitionTime: 2019-03-12T07:50:41Z
  message: Deployment "target-deploy-abc" has been successfully patched.
  lastError: 
  startTime: 
  endTime: 
  retries: 
- name: patch-cvr
  status: CStor volume replica "cvr-1" has been successfully patched.
  lastTransitionTime: 2019-03-12T07:59:40Z
  message: 
  lastError: 
  startTime: 
  endTime: 
  retries: 
status:
  DesiredCount: 2
  ActualCount: 1
  FailedCount: 1
  resource:
    name: pv-1
    kind: Persistent Volume
    apiVersion: v1
    namespace: default
    status: Healthy
    lastTransitionTime: 2019-03-12T07:59:40Z
    message: PV "pv-1" is healthy.
    preState:
    - name: cvr-1
      kind: CStorVolumeReplica
      apiVersion: openebs.io/v1alpha1
      namespace: openebs
      status: Healthy
      lastTransitionTime: 2019-03-12T06:59:46Z
      message: CStor volume Replica "cvr-1" is healthy.
    - name: target-deploy-abc
      kind: Deployment
      apiVersion: extensions/v1beta1
      namespace: openebs
      status: Running
      lastTransitionTime: 2019-03-12T06:59:46Z
      message: Deployment "target-deploy-abc" is running.
    postState:
    - name: cvr-1
      kind: CStorVolumeReplica
      apiVersion: openebs.io/v1alpha1
      namespace: openebs
      status: Healthy
      lastTransitionTime: 2019-03-12T06:59:46Z
      message: CStor volume Replica "cvr-1" is healthy after upgrade.
    - name: target-deploy-abc
      kind: Deployment
      apiVersion: extensions/v1beta1
      namespace: openebs
      status: Running
      lastTransitionTime: 2019-03-12T06:59:46Z
      message: Deployment "target-deploy-abc" is running.
    
```

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Checklist:**
- [ ] Fixes #<issue number>
- [ ] Labelled this PR & related issue with `documentation` tag
- [ ] PR messages has document related information
- [ ] Labelled this PR & related issue with `breaking-changes` tag
- [ ] PR messages has breaking changes related information
- [ ] Labelled this PR & related issue with `requires-upgrade` tag
- [ ] PR messages has upgrade related information
- [ ] Commit has unit tests
- [ ] Commit has integration tests

### CODE REVIEW

File: pkg/apis/openebs.io/upgrade/v1alpha1/doc.go [hunk 1]
```diff
@@ -0,0 +1,21 @@
+/*
```
Comment (AmitKumarDas) [inline pkg/apis/openebs.io/upgrade/v1alpha1/doc.go]: Please refer to https://github.com/openebs/maya/tree/master/pkg/apis/openebs.io/catalog/v1alpha1
for doc.go, register.go & other xyz_file.go

File: pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+	// TODO: Add task-executor config spec here
+
+	// BaseVersion is the current version of openebs
+	BaseVersion string `json:"baseVersion"`
+	// TargetVersion is the version to which openebs
+	// components needs to be upgraded
+	TargetVersion string `json:"targetVersion"`
+	// ResourceName contains comma separated name of
+	// the resources which needs to be upgraded
+	ResourceName string `json:"resourceName"`
```
Comment (AmitKumarDas) [inline pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go]: Why is this required? If the entire config will be embedded that should provide the details.

File: pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+// +resource:path=upgrade
+
+// UpgradeResult forms the desired specification of
+// upgrade task's result that should be available for
+// all the upgrade tasks
+type UpgradeResult struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   UpgradeResultSpec   `json:"spec"`
```
Comment (AmitKumarDas) [inline pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go]: We would like to add Config as a direct property. This should hold the entire config pushed to worker upgrade job.

File: pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+	Status UpgradeResultStatus `json:"status"`
+}
+
+// UpgradeResultSpec is the specifications of
+// upgrade result
+type UpgradeResultSpec struct {
+	// TODO: Add task-executor config spec here
```
Comment (AmitKumarDas) [inline pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go]: Let us not add config inside spec. Let it be outside.
Alternatively, the entire spec can be the config. Let us not introduce any new properties in the spec.
We might want to rename spec to config also.
Please think about all these options.

File: pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+	TargetVersion string `json:"targetVersion"`
+	// ResourceName contains comma separated name of
+	// the resources which needs to be upgraded
+	ResourceName string `json:"resourceName"`
+}
+
+// UpgradeResultStatus represents the current
+// status of UpgradeTaskResult
+type UpgradeResultStatus struct {
+	// TotalTaskCount is the total no tasks that
```
Comment (AmitKumarDas) [inline pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go]: We might want to rethink the status when worker upgrade job can upgrade more than one volumes.
e.g. unit of upgrade is volume
there are 3 volumes to be upgraded.
It makes sense to let the worker job invoke, upgrade these 3 volumes one-by-one by executing the same cas-template-to-upgrade-volume three times, and each time different runtime values/arguments are provided to castemplate.

With above scenario in mind, we have to come up with the status schema in such a way that every resource's task is accounted for.

File: pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+	// needs to be executed to upgrade a particular
+	// openebs component to desired version
+	TotalTaskCount int `json:"totalTaskCount"`
+	// CompletedTaskCount represents the no of tasks
+	// that has successfully completed
+	CompletedTaskCount int `json:"completedTaskCount"`
+	// FailedTaskCount represents the no of tasks
+	// that has failed
+	FailedTaskCount int `json:"failedTaskCount"`
```
Comment (AmitKumarDas) [inline pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go]: We might want to include `Message []string`

File: pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+	// Tasks represents the list of tasks
+	Tasks []Task `json:"tasks"`
```
Comment (AmitKumarDas) [inline pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go]: Can we make use of `Resources []Resource` instead of task.
We are promoting resource versus task.
A resource in then will consist of sub-tasks i.e. runtasks.

File: pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+// Task represents an upgrade task and its subtasks details
+type Task struct {
```
Comment (AmitKumarDas) [inline pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go]: If we think Resource is better than Task, then can we include Name, Kind, APIVersion, Namespace, BeforeUpgrade & AfterUpgrade state of the resource, etc

File: pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+	// Name of the task
+	Name string `json:"name"`
+	// TotalSubTaskCount is the no of subtasks
+	// i.e. runtasks that needs to be executed
+	// to complete a task
+	TotalSubTaskCount int `json:"totalSubTaskCount"`
```
Comment (AmitKumarDas) [inline pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go]: I would prefer to include every minute detail of the runtask. in other words i would go for below:
```go
type Resource struct {
  Name   string
  APIVersion  string
  Namespace  string
  BeforeUpgrade  ResourceState
  AfterUpgrade     ResourceState
  Tasks   []Task
}

type Task struct {
  Name             string
  Message        string
  LastError         error
  TimeTaken   duration
  StartTime     time
  EndTime      time
  retries          int
  errorCount   int
}
```

Comment (AmitKumarDas) [reply pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go]: This might require a change to core castengine execution workflow. However, the most important thing here is upgrade should work without issues.

File: pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go [hunk 10]
```diff
@@ -20,76 +20,123 @@ import (
...
+	Name string `json:"name"`
+	// Kind is the type of resource i.e.
+	// cvr, deployment, ..
+	Kind string `json:"kind"`
+	// APIVersion of the resource
+	APIVersion string `json:"apiVersion"`
+	// Namespace of the resource
+	Namespace string `json:"namespace"`
+	// Status of the resource
+	Status string `json:"status"`
```
Comment (sagarkrsd) [inline pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go]: Does this status also needs to be a type struct with more details ?

File: pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go (hunk 7 above)
Comment (sagarkrsd) [reply pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go]: Have added Resources []Resource and under it []task for runtasks for a particular resource upgrade.

File: pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+	// that has failed to upgrade
+	FailedResources int `json:"failedResources"`
+	// ResourceList is the list of resources that needs to
+	// be upgraded
+	ResourceList []UpgradeResource `json:"resourceList"`
+}
+
+// UpgradeResource represents a resource that needs to
+// be upgraded to a desired version
+type UpgradeResource struct {
```
Comment (AmitKumarDas) [inline pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go]: Can we have below structure ?
```go
type UpgradeResource struct {
  Resource
  PreState              ResourceState  `json:"preState"`
  PostState            ResourceState  `json:"postState"`
  SubResources    []UpgradeSubResource  `json:"subResources"`
}

type UpgradeSubResource struct {
  Resource
  PreState      ResourceState  `json:"preState"`
  PostState    ResourceState  `json:"postState"`
}

type Resource struct {
  Name            string   `json:"name"`
  Kind              string   `json:"kind"`
  Namespace  string   `json:"namespace"`
  APIVersion    string   `json:"apiVersion"`
}

type ResourceState struct {
  Status                       string `json:"status"`
  LastTransitionTime   metav1.Time `json:"lastTransitionTime"`
  Message                   string  `json:"message"`
}
```

File: pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+// +resource:path=upgrade
+
+// UpgradeResult contains the desired specifications of an
+// upgrade result
+type UpgradeResult struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Config UpgradeResultConfig `json:"config"`
+	Status UpgradeResultStatus `json:"status"`
```
Comment (AmitKumarDas) [inline pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go]: Will it be good to have Tasks here?
```go
type UpgradeResult struct {
  ...
  ...
  Config  UpgradeResultConfig  `json:"config"`
  Tasks   UpgradeResultTask    `json:"tasks"`
  Status  UpgradeResultStatus  `json:"status"`
}

File: pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go (hunk 10 above)
Comment (AmitKumarDas) [reply pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go]: not needed as of now

File: pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go (hunk 12 above)
Comment (sagarkrsd) [reply pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go]: I believe tasks should be part of the resources itself since it would be helpful to refer to the runtasks of a particular resource say runtasks of pool A or B.

Comment (AmitKumarDas) [reply pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go]: @sagarkrsd it might be good. it seems to be good as we can cross refer easily.
However, i wanted to delink the logic part from the status part.
e.g. a task can have some retries, errorCount, successCount, time taken & so on. However this are logic specific.
The resources & subresources should be restricted to their schema i.e. display preState, postState.
The resources & sub-resources should show their CR state only. we should not set the status of a resource with task executor message.

Comment (sagarkrsd) [reply pkg/apis/openebs.io/upgrade/v1alpha1/upgraderesult.go]: Thanks @AmitKumarDas , will do that.
